### ISSUE TITLE
Integer stream parsing

### ISSUE THREAD

**Issue by barakugav** [mohanson/leb128#12]
I have a long sequence of bytes encoding multiple messages, each message is encoded as its length (using leb128) and then the message body itself.
With the current API, I can not simply pass a too large array to the library to parse the first bytes as length, so I do it manually:
```python
# returns a tuple of the message bytes and the number of bytes consumed (including the leb128 encoding and the message body)
def decode_length_and_message(bytes: bytes) -> tuple[bytes, int]:
    length_bytes_num = 1
    while bytes[length_bytes_num - 1] >= 0x80:
        if length_bytes_num >= len(bytes):
            raise Exception("length bytes num is too large")
        length_bytes_num += 1
    if length_bytes_num > 10:
        raise Exception("length bytes num is too large")

    msg_len = leb128.u.decode(bytes[:length_bytes_num])
    message_bytes = bytes[length_bytes_num : length_bytes_num + msg_len]

    return message_bytes, length_bytes_num + msg_len
```

It would be great if the library could provide a function the decode integer _without consuming the whole_ buffer.
Such function will highly simplify my code. Possible implementation:
```python
def decode_leb128(bytes: bytes) -> tuple[int, int]:
    length_bytes_num = 1
    while bytes[length_bytes_num - 1] >= 0x80:
        if length_bytes_num >= len(bytes):
            raise Exception("length bytes num is too large")
        length_bytes_num += 1
    if length_bytes_num > 10:
        raise Exception("length bytes num is too large")

    val = leb128.u.decode(bytes[:length_bytes_num])
    return val, length_bytes_num
```
I think this is fairly common use case, what do you think?

Thanks for the support! This library is awesome and useful :)

**Comment by mohanson** [comment]
Stream parsing is already supported. You can use the following API:

```py
leb128.u.decode_reader(stream)
```

The function will return the number of bytes read and the integer.

**Comment by barakugav** [comment]
thanks!

**Comment by barakugav** [comment]
Is it possible to throw a unique exception in case the stream is empty? currently `ord(r.read(1))` fails.
Maybe just return `None`

**Comment by mohanson** [comment]
I can't do that because I don't know what error will occur when the stream is empty. You should handle this exception yourself since you are the owner of the stream.

**Comment by barakugav** [comment]
The strean itself doesn't raise an exception when its empty, rather its returns an empty byte string, and the conversion in the library will fail

**Comment by mohanson** [comment]
I just released v1.0.8. When the reader returns an empty string it will now trigger an EOF error, which should help you.

https://github.com/mohanson/leb128/commit/d53bf789601d12ad9bf6c3dee9bda1c4cb38bff5

**Comment by barakugav** [comment]
Thanks! I appreciate the help :)
